### PR TITLE
Transition GPU support to weak-deps and extensions

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

**Comment by simonbyrne** [comment]
ERROR: LoadError: MethodError: no method matching version()
```
Closest candidates are:
  version(::CUDA.HostKernel) at ~/.cache/julia-buildkite-plugin/depots/03f2e4c8-d\
151-49cd-bebb-f0893982b35a/packages/CUDA/ZdCxS/src/compiler/execution.jl:235
Stacktrace:
 [1] top-level scope
   @ /var/lib/buildkite-agent/builds/gpuci-12/julialang/mpi-dot-jl/test/runtests.\
jl:9
 [2] include(fname::String)
   @ Base.MainInclude ./client.jl:476
 [3] top-level scope
   @ none:6
in expression starting at /var/lib/buildkite-agent/builds/gpuci-12/julialang/mpi-\
dot-jl/test/runtests.jl:7
```



ERROR: LoadError: MethodError: no method matching version()
--


in expression starting at /var/lib/buildkite-agent/builds/gpuci-12/julialang/mpi-dot-jl/test/runtests.jl:7

**Comment by luraess** [comment]
`CUDA.versioninfo()` exists for sure now.

File: Project.toml [hunk 1]
```diff
@@ -16,7 +16,21 @@ Requires = "ae029012-a4dd-5104-9daa-d747884805df"
...
+AMDGPU = "21141c5a-9bdb-4563-92ae-f87d6854732e"
+CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
+
+[weakdeps]
+AMDGPU = "21141c5a-9bdb-4563-92ae-f87d6854732e"
+CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
+
+[extensions]
+AMDGPUExt = "AMDGPU"
+CUDAExt = "CUDAExt"
```
Comment (vchuravy) [inline Project.toml:29]:
```suggestion
CUDAExt = "CUDA"
```

File: ext/AMDGPUExt.jl [hunk 2]
```diff
@@ -1,4 +1,9 @@
-import .AMDGPU
+module AMDGPUExt
+
+import MPI
+isdefined(Base, :get_extenstion) ? (import AMDGPU) : (import ..AMDGPU)
```
Comment (vchuravy) [inline ext/AMDGPUExt.jl:4]:
```suggestion
isdefined(Base, :get_extension) ? (import AMDGPU) : (import ..AMDGPU)
```

File: ext/CUDAExt.jl [hunk 3]
```diff
@@ -1,4 +1,8 @@
-import .CUDA
+module CUDAExt
+
+import MPI 
+isdefined(Base, :get_extenstion) ? (import CUDA) : (import ..CUDA)
```
Comment (vchuravy) [inline ext/CUDAExt.jl:4]:
```suggestion
isdefined(Base, :get_extension) ? (import CUDA) : (import ..CUDA)
```

File: src/MPI.jl [hunk 4]
```diff
@@ -80,6 +79,10 @@ include("misc.jl")
 
 include("deprecated.jl")
 
+if !isdefined(Base, :get_extension)
```
Comment (sloede) [inline src/MPI.jl:82]: Out of curiosity (and such that I can learn from it): Why isn't this guarded by `@static` as well, e.g.
```julia
@static if !isdefined(Base, :get_extension)
```
?

Comment (giordano) [reply src/MPI.jl:82]: `@static` doesn't do much for top-level expressions, they're always evaluated at precompile time anyway.

Comment (sloede) [reply src/MPI.jl:82]: Ah, thanks a lot for the clarification!

**Pull request merged**